Upgrade SvelteKit

AndrewLester opened the pull request:
## Status:
:rocket: Ready

## Description

Upgrades SvelteKit version to fix prefetching.

Fixes: #55 
Fixes: #56 

## Screenshots

Ignore the errors, I can't connect to the API.

https://user-images.githubusercontent.com/23221268/158862741-293d1fc6-3be3-4b5d-aea4-f6e98c166d41.mov

AndrewLester committed:
Upgrade SvelteKit to fix prefetching

SirajChokshi reviewed (changes requested):
Thanks for maintaining this.

The pre-fetching is great, but it seems like the mobile navigation toggle is broken (https://deploy-preview-57--h4iuiuc.netlify.app/)

AndrewLester commented:
> Thanks for maintaining this.
> 
> The pre-fetching is great, but it seems like the mobile navigation toggle is broken (https://deploy-preview-57--h4iuiuc.netlify.app/)

That's unfortunate, the preview deployments really come in handy. After checking the console, I think some script is erroring out and it's crashing Svelte before it can register the button on:click listeners. Here's a screenshot of the error's location in the script, looks like it might be ackee?
<img width="306" alt="image" src="https://user-images.githubusercontent.com/23221268/159630609-ac890677-7d20-4327-90e0-9ac6d2d58742.png">

Here's a screenshot of a potential root cause (note the `{}.CLIENT_ACKEE_URL`, which yields undefined):
<img width="994" alt="image" src="https://user-images.githubusercontent.com/23221268/159633776-18011b8d-2e8b-4cf4-8001-968264ef983d.png">


You might have a better idea of why this is happening, I'm not entirely sure though. Maybe the Ackee urls aren't in the preview?

arpanlaha commented:
> > Thanks for maintaining this.
> > The pre-fetching is great, but it seems like the mobile navigation toggle is broken (https://deploy-preview-57--h4iuiuc.netlify.app/)
> 
> That's unfortunate, the preview deployments really come in handy. After checking the console, I think some script is erroring out and it's crashing Svelte before it can register the button on:click listeners. Here's a screenshot of the error's location in the script, looks like it might be ackee? <img alt="image" width="306" src="https://user-images.githubusercontent.com/23221268/159630609-ac890677-7d20-4327-90e0-9ac6d2d58742.png">
> 
> Here's a screenshot of a potential root cause (note the `{}.CLIENT_ACKEE_URL`, which yields undefined): <img alt="image" width="994" src="https://user-images.githubusercontent.com/23221268/159633776-18011b8d-2e8b-4cf4-8001-968264ef983d.png">
> 
> You might have a better idea of why this is happening, I'm not entirely sure though. Maybe the Ackee urls aren't in the preview?

Ackee was me, I'll take a look soon (probably the weekend).

AndrewLester commented:
Coming back to this, it might just be better to upgrade SvelteKit to its most recent version since a lot has changed. The SvelteKit docs really only describe its most recent version, so it will be easier to maintain with a newer version in the codebase. I've recently done a SvelteKit upgrade, so I know generally what needs to be changed in the codebase and I could handle it in this PR. Thoughts?

arpanlaha commented:
> Coming back to this, it might just be better to upgrade SvelteKit to its most recent version since a lot has changed. The SvelteKit docs really only describe its most recent version, so it will be easier to maintain with a newer version in the codebase. I've recently done a SvelteKit upgrade, so I know generally what needs to be changed in the codebase and I could handle it in this PR. Thoughts?

Sure go for it, sorry I forgot to follow up earlier. Let me know once you've bumped to a more up-to-date version and I'll take a look again.

AndrewLester committed:
Update SvelteKit to most recent version

AndrewLester committed:
Update general config files

AndrewLester committed:
Replace globals.d.ts and separate app css and html

AndrewLester committed:
Rename HTMLProps to HTMLAttributes in additionall svelte JSX

AndrewLester committed:
Fix hooks import

AndrewLester committed:
Don't use onClick prop instead forward click event

AndrewLester committed:
Refactor imports to reflect aliases and fix many types

AndrewLester committed:
Make work with us swap more explicit

AndrewLester committed:
Refactor load functions to use types

AndrewLester committed:
Fix some template type errors by using new svelte template tags

AndrewLester commented:
This PR unfortunately became quite large because of the upgrade. Here's a rundown of why I made some changes that increased the files changed by a lot:
1. Removing the custom import alias for `$lib` (was pointing at `src`). I removed this because it was messing up SvelteKit's new generated types system. In doing so, I added new import aliases which resulted in almost every one changing.
2. More explicit typing of endpoint and page load functions. Explicitly typing these functions allows SvelteKit's generated types to work when necessary, meaning page params are typed. I wanted to make this consistent so I typed every one, changing all of these functions to arrow functions in the process. Since not all of them made use of SvelteKit's generated types, for some I just imported their type from the normal SvelteKit type.
3. Changed type definitions from `interface` to `type`. I made this change because of an issue I came across before when returning `interface` defined types in endpoint functions. For some reason, SvelteKit's type system doesn't think `interface` defined types are JSON compatible. Swapping to `type` fixes this issue. [This TS playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAUgygeQHIDUCGAbArtAvFAZ2ACcBLAOwHMoAfWRJBAIwCsIBjYAbgCgeLgEYgDM07aPGTM2nKAG8eASADaAawggAXIRIVKAXW2TUmHLwC+fAUNHioASQC2aShHvlBIsdAVQ-hYnZtIjIqCz5QSAdnVwAVcDx5Hn8AoJ1QynCedgB7ciIoay9xaSMGaQ5gKHw5VO0AckEieqhzKDQCaJc3DxtvXmy8gsiIUvopVkrq+TqoRohm1vbOp274yC4gA) displays the issue.

Also, SvelteKit now requires node 16 I believe. Netlify might need to upgrade? I've never used it though so I'm not entirely sure if I'm reading the logs right.

Lastly, I'm going to make this a draft until I get the contentful API keys and can test locally.

AndrewLester committed:
Add prepare script and fix formatting

AndrewLester committed:
Set prerender default to true

SirajChokshi commented:
Thanks for all the work on this @AndrewLester 🚀 

I think @arpanlaha has access to the Netlify setup, there should be an option to upgrade to Node 16 LTS.

AndrewLester committed:
Replace navigation end event with afterNavigate lifecycle function

AndrewLester committed:
Remove some build steps by integrating 404 page and minification

AndrewLester committed:
Use any types without reducing safety elsewhere

AndrewLester committed:
Upgrade eslint plugin

AndrewLester committed:
Split build script into post build

arpanlaha commented:
I upgraded Node, here's the deploy preview: https://deploy-preview-57--h4iuiuc.netlify.app/ (doesn't display on CI since I went through the netlify thing itself)

arpanlaha commented:
I added the Ackee env vars to the preview Netlify and we still get a network error (Ackee needs to specify a specific domain which we can't do with preview deploys) but IIUC it isn't blocking the onclick handling now. @AndrewLester could you confirm?

AndrewLester committed:
Only run ackee if variables exist

AndrewLester committed:
Fix refetch of layout

arpanlaha commented on .eslintrc.cjs:
```diff
@@ -1,11 +1,16 @@
+  extends: [
+    "eslint:recommended",
+    "plugin:@typescript-eslint/recommended",
+    "prettier",
```
I personally like to keep linting and formatting separate - I'm not going to be super hands-on moving forward and no longer "own" this repo so this isn't blocking feedback but I wanted to mention it.

AndrewLester replied:
To be honest, I didn't know what this setting did when I added it. I just copied it from the SvelteKit template. After some Googling, it seems as though placing `prettier` in this file enforces a linting and formatting separation, based on [this docs page](https://prettier.io/docs/en/integrating-with-linters.html) and because this PR adds `eslint-config-prettier` to the package.json.

arpanlaha commented on .eslintrc.cjs:
```diff
@@ -14,7 +19,7 @@ module.exports = {
   parserOptions: {
     sourceType: "module",
-    ecmaVersion: 2019,
+    ecmaVersion: 2020,
```
Why this change?

AndrewLester replied:
Just took this from the new template: https://github.com/sveltejs/kit/blob/master/packages/create-svelte/shared/%2Beslint%2Bprettier%2Btypescript/.eslintrc.cjs

arpanlaha commented on src/app.css:
```diff
@@ -0,0 +1,232 @@
+:root {
+  --blue: #155da1;
+  --blue-light: #3e87cd;
+  --blue-lighter: #b4d8fa;
+  --blue-dark: #0e4e8a;
+  --blue-darker: #0d3f6e;
+  --gray-lighter: #ebeef2;
+  --gray-light: #8b9199;
+  --text-dark: #031425;
+
+  --header-width: min(1440px, 90vw);
+  --content-width: min(1200px, 85vw);
+  --long-form-width: min(900px, 80vw);
+
+  --animation: 0.2s ease;
```
I intentially inlined global CSS in app.html - is there a reason you separated it out? I was under the impression that inlining global styles was a good idea to cut down on a network request, but I'm not an expert on this.

AndrewLester replied:
I separated it out because I've seen this pattern a lot for SvelteKit, and it's how [the template](https://sveltekit.new) is setup. It does add a network request, but this can be avoided with [inlineStyleThreshold](https://kit.svelte.dev/docs/configuration#inlinestylethreshold). That said, the file is a bit large to set an inline threshold for in my opinion, so in the future I might try to shrink it by putting some of these styles into components and removing them from here, but for now I'll just revert this back to app.html. I'll still keep the inline style threshold at 1024 bytes for any small files currently.

The only downside to inlining css manually is that Vite's dev server doesn't apply HMR to edits. Also, I think it's a bit easier to read when separated into its own file.

AndrewLester committed:
Fix svelte:head swapping

arpanlaha commented on src/routes/__layout.svelte:
```diff
@@ -1,33 +1,37 @@
 <script lang="ts">
   import * as ackeeTracker from "ackee-tracker";
   import { beforeUpdate } from "svelte";
+  import "../app.css";
```
If we do go ahead with not inlining global CSS in `app.html`, wouldn't it be more idomatic to add it to the style tag in this file?

AndrewLester replied:
Addressed in the other comment, but the short answer is it's personal preference I guess? If we put it in the style tag we'd need to add `:global` prefixes to a bunch of the selectors. But, it would be easier to find in there.

arpanlaha commented on src/utils/api.ts:
```diff
@@ -25,27 +25,25 @@ export class ContentWrapper {
 
     const [entries, schema] = await Promise.all([
-      this.client.getEntries({
+      this.client!.getEntries({
```
Looking back at this, instead of doing the `!` assertion it would be better for `this.build` to return `createClient(...)` and change the beginning lines of this function to:

```
if (this.client === undefined) {
      this.client = await this.build();
}
```

From there, TypeScript will be able to figure out that `this.client` exists without the `!` assertion.

AndrewLester replied:
I'll go with this.

arpanlaha commented on src/utils/projects.ts:
```diff
@@ -1,15 +1,15 @@
+export type SemesterProjects = {
   featured?: Project;
   projects: Project[];
-}
+};
 
-export interface ProjectsInfo {
+export type ProjectsInfo = {
   projectMap: Record<string, SemesterProjects>;
   semesters: string[];
-}
+};
```
Another personal preference for TS - I like to default to `interface` and only use `type` if I'm making use of union types or generics (i.e. `type Status = 'good' | 'bad'` or `type Result = Promise<...>`.

This is also nonblocking feedback

AndrewLester replied:
This is a difficult one because there are tradeoffs to both sides since SvelteKit is a bit weird. With interfaces, another type param must be added whenever these interfaces are returned from an endpoint function. See here: https://github.com/sveltejs/kit/issues/5468#issuecomment-1182231341
Also, the RequestHandler type must be imported from the SvelteKit generated types which isn't as much of an issue.

Using a type, this is not necessary. I personally think it looks cleaner when it's implicit, but I'm not entirely sure if this is enough to justify a change. Here's a TS playground that shows roughly why interface typed objects can't be returned from endpoint functions (without the steps in the comment above): [playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAUgygeQHIDUCGAbArtAvFAZ2ACcBLAOwHMoAfWRJBAIwCsIBjYAbgCgeLgEYgDM07aPGTM2nKAG8eASADaAawggAXIRIVKAXW2TUmHLwC+fAUNHioASQC2aShHvlBIsdAVQ-hYnZtIjIqCz5QSAdnVwAVcDx5Hn8AoJ1QynCedgB7ciIoay9xaSMGaQ5gKHw5VO0AckEieqhzKDQCaJc3DxtvXmy8gsiIUvopVkrq+TqoRohm1vbOp274yC4gA)

arpanlaha replied:
Ah gotcha, since it's a Svelte issue sgtm

arpanlaha commented on src/utils/schema.ts:
```diff
@@ -1,17 +1,21 @@
-}
+};
+
+export type TestimonialMember = Member & {
+  testimonial: string;
+};
```
Going off of the `interface` vs. `type` comment above, this could be done as `export interface TestimonialMember extends Member {...`

AndrewLester replied:
Will update this to reflect a decision on the above.

arpanlaha commented on src/utils/useViewportAction.ts:
```diff
@@ -12,15 +12,15 @@ function ensureIntersectionObserver() {
 
-export default function viewport(element: HTMLSpanElement) {
+export default function viewport(element: HTMLElement) {
   if (element) {
     ensureIntersectionObserver();
 
-    intersectionObserver.observe(element);
+    intersectionObserver!.observe(element);
```
You could use the same approach as the API client to avoid `!` assertions here

arpanlaha reviewed (commented):
Thanks for doing this, some feedback that is either relatively small or non-blocking:

AndrewLester committed:
Revert separation of css from app.html and create root layout

AndrewLester committed:
Fix intersection observer and api client typing with same structure

AndrewLester committed:
Fix format

arpanlaha reviewed (approved):
Sweet

AndrewLester committed:
Upgrade SvelteKit again

AndrewLester committed:
Set preview port to 3000

AndrewLester commented:
One last change I made was to separate the layout file into two, `__layout-root.svelte` and `__layout@root.svelte`. The root layout file can be inherited directly by any page to "reset" the layout without missing necessary stuff like ackee (and global CSS in the future). The default layout file that inherits from root has all the actual stuff for page layout like the nav and `<main>`.

AndrewLester committed:
Use SvelteKit dev variable for prod check

AndrewLester committed:
Fix 404 page not showing up

SirajChokshi reviewed (approved):
Great work 💯

AndrewLester committed:
Update readme for SvelteKit clarification

AndrewLester committed:
Remove ambient ts

AndrewLester committed:
Revert "Remove ambient ts"

This reverts commit 7ee7e5ac48d101b3843ab6ae8a4fcb53f3381cd6.
